### PR TITLE
feat: implement copyright update

### LINKED ISSUE #27: Add automation for updating copyright information
CRS has the year written on its copyright information on every rules file. So each new year we need to be able to update it easily.

**Requirements**
- go over each rule file and update the year. An example could be:
```
# OWASP ModSecurity Core Rule Set ver.4.0.0-rc1
# Copyright (c) 2006-2020 Trustwave and contributors. All rights reserved.
# Copyright (c) 2021-2022 Core Rule Set project. All rights reserved.
```
Should change, next year to:
```
# OWASP ModSecurity Core Rule Set ver.4.0.0-rc1
# Copyright (c) 2006-2020 Trustwave and contributors. All rights reserved.
# Copyright (c) 2021-2023 Core Rule Set project. All rights reserved.
```

**Output**
- a new command (or option) to do this. If we think that this is part of a list of _chores_ we will be doing for the CRS, the `crs-toolchain chore copyright-update` might work?

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Updates copyright and crs version in conf files.

Closes #27 and #28.

### CODE REVIEW

**Comment by fzipi** [comment]
Ugh, this should be a draft. Bad button pressed :/

File: chore/copyright_update.go
Comment (theseion) [inline chore/copyright_update.go:1]: I'd prefer the new command to be named `update-copyright`. That form matches the other commands better, like `renumber-tests`.

File: chore/copyright_update.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/coreruleset/crs-toolchain/context"
+	"github.com/coreruleset/crs-toolchain/regex"
+)
+
+var logger = log.With().Str("component", "copyright-update").Logger()
+
+// CopyrightUpdate updates the copyright portion on the rules files to the provided year and version.
```
Comment (theseion) [inline chore/copyright_update.go:20]:
```suggestion
// CopyrightUpdate updates the copyright portion of the rules files to the provided year and version.
```

File: chore/copyright_update.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		return err
+	}
+
+	output, err := updateRules(version, year, contents)
+	if err != nil {
+		return err
+	}
+
+	err = os.WriteFile(filePath, output, fs.ModePerm)
+	if err != nil {
```
Comment (theseion) [inline chore/copyright_update.go:59]: No need for the nil check here.

```suggestion
	return err
```

File: chore/copyright_update.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// Ideally we have support in the future for a proper parser file so we can use that to change it
+// in a more elegant way. Right now we just match strings.
+func updateRules(version string, year string, contents []byte) ([]byte, error) {
+	scanner := bufio.NewScanner(bytes.NewReader(contents))
+	scanner.Split(bufio.ScanLines)
+	output := new(bytes.Buffer)
+	writer := bufio.NewWriter(output)
+	for scanner.Scan() {
+		line := scanner.Text()
+		replaceVersion := fmt.Sprintf("${1}%s", version)
```
Comment (theseion) [inline chore/copyright_update.go:74]: Move this constant out of the loop.

File: chore/copyright_update.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		line = regex.CRSVersionRegex.ReplaceAllString(line, replaceVersion)
+		replaceYear := fmt.Sprintf("${1}%s${3}", year)
```
Comment (theseion) [inline chore/copyright_update.go:76]: Move this constant out of the loop.

File: chore/copyright_update.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+			return nil, err
+		}
+		if _, err := writer.WriteRune('\n'); err != nil {
+			return nil, err
+		}
+	}
+
+	writer.Flush()
+	outputBytes := output.Bytes()
+	// if the file was empty, we don't want to remove the newline character
```
Comment (theseion) [inline chore/copyright_update.go:89]:
```suggestion
	// if the file was empty, we didn't change anything and we're done
```

File: cmd/chore_copyright_update.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func init() {
+	buildChoreCopyrightUpdateCommand()
+}
+
+func createChoreCopyrightUpdateCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:   "copyright-update",
+		Short: "Updates the copyright on every rule file",
```
Comment (theseion) [inline cmd/chore_copyright_update.go:29]:
```suggestion
		Short: "Updates the copyright in setup, example setup, and rule files",
```

File: cmd/chore_copyright_update.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			rootContext := context.New(rootValues.workingDirectory.String(), rootValues.configurationFileName.String())
+			chore.CopyrightUpdate(rootContext, copyrightVariables.Version, copyrightVariables.Year)
+		},
+	}
+}
+
+func buildChoreCopyrightUpdateCommand() {
+	choreCmd.AddCommand(choreCopyrightUpdateCmd)
+	choreCopyrightUpdateCmd.Flags().StringVarP(&copyrightVariables.Year, "year", "y", strconv.Itoa(time.Now().Year()), "Year, example: 2023. Default: current year.")
```
Comment (theseion) [inline cmd/chore_copyright_update.go:39]:
```suggestion
	choreCopyrightUpdateCmd.Flags().StringVarP(&copyrightVariables.Year, "year", "y", strconv.Itoa(time.Now().Year()), "Four digit year")
```

File: cmd/chore_copyright_update.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	choreCopyrightUpdateCmd.Flags().StringVarP(&copyrightVariables.Version, "version", "v", "CRS v10.0.1", "Add this text as the version to the file.")
```
Comment (theseion) [inline cmd/chore_copyright_update.go:40]:
```suggestion
	choreCopyrightUpdateCmd.Flags().StringVarP(&copyrightVariables.Version, "version", "v", "CRS v10.0.1", "Add this text as the version to the file.")
		choreCopyrightUpdateCmd.Flags().Lookup("version").DefValue = "None"
```

Comment (theseion) [inline cmd/chore_copyright_update.go:40]: The version must be required, or you need to figure out the current version automatically.

File: cmd/chore_copyright_update_test.go
Comment (theseion) [inline cmd/chore_copyright_update_test.go:1]: Add a test for `setup.conf` and `setup.conf.example`, if you want to update those.

File: regex/definitions.go [hunk 9]
```diff
@@ -76,3 +76,11 @@ var TestTitleRegex = regexp.MustCompile(`(.*test_title:).*$`)
...
 // The matched reference name will be captured in group 1.
 var DefinitionReferenceRegex = regexp.MustCompile(`{{([a-zA-Z0-9-_]+)}}`)
+
+// CRSVersionRegex matches the version contained on every rules file.
+// The version declared on the file is captured in group 1.
+var CRSVersionRegex = regexp.MustCompile(`^(# OWASP ModSecurity Core Rule Set ver\.)(.+)$`)
+
+// CRSCopyrightYear matches the year of the copyright text on every rules file.
+// The "to" year is captured in group 1.
+var CRSCopyrightYear = regexp.MustCompile(`^(# Copyright \(c\) 2021-)(\d{4})( Core Rule Set project. All rights reserved.)$`)
```
Comment (theseion) [inline regex/definitions.go:86]:
```suggestion
var CRSCopyrightYearRegex = regexp.MustCompile(`^(# Copyright \(c\) 2021-)(\d{4})( Core Rule Set project. All rights reserved.)$`)
```

File: regex/definitions.go [hunk 10]
```diff
@@ -76,3 +76,11 @@ var TestTitleRegex = regexp.MustCompile(`(.*test_title:).*$`)
 // DefinitionReferenceRegex matches any reference to a definition.
 // The matched reference name will be captured in group 1.
 var DefinitionReferenceRegex = regexp.MustCompile(`{{([a-zA-Z0-9-_]+)}}`)
+
+// CRSVersionRegex matches the version contained on every rules file.
+// The version declared on the file is captured in group 1.
+var CRSVersionRegex = regexp.MustCompile(`^(# OWASP ModSecurity Core Rule Set ver\.)(.+)$`)
+
+// CRSCopyrightYear matches the year of the copyright text on every rules file.
+// The "to" year is captured in group 1.
```
Comment (theseion) [inline regex/definitions.go:85]:
```suggestion
// The matched end year of the copyright year range will be captured in group 1.
```

File: regex/definitions.go [hunk 11]
```diff
@@ -76,3 +76,11 @@ var TestTitleRegex = regexp.MustCompile(`(.*test_title:).*$`)
 // DefinitionReferenceRegex matches any reference to a definition.
 // The matched reference name will be captured in group 1.
 var DefinitionReferenceRegex = regexp.MustCompile(`{{([a-zA-Z0-9-_]+)}}`)
+
+// CRSVersionRegex matches the version contained on every rules file.
+// The version declared on the file is captured in group 1.
+var CRSVersionRegex = regexp.MustCompile(`^(# OWASP ModSecurity Core Rule Set ver\.)(.+)$`)
+
+// CRSCopyrightYear matches the year of the copyright text on every rules file.
```
Comment (theseion) [inline regex/definitions.go:84]:
```suggestion
// CRSCopyrightYearRegex matches the version and year range of the copyright text in setup,
// setup example, and rule files.
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: chore/copyright_update.go (hunk 2 above)
Comment (fzipi) [reply chore/copyright_update.go:59]: Done.

File: chore/copyright_update.go (hunk 3 above)
Comment (fzipi) [reply chore/copyright_update.go:74]: Done.

File: chore/copyright_update.go (hunk 4 above)
Comment (fzipi) [reply chore/copyright_update.go:76]: Moved.

File: cmd/chore_copyright_update.go (hunk 8 above)
Comment (fzipi) [reply cmd/chore_copyright_update.go:40]: We can add both. Let me think about something.

Comment (fzipi) [reply cmd/chore_copyright_update.go:40]: Right now I think the easy way is to require. We can add more things like reading the git repo in the future.

File: cmd/chore_copyright_update_test.go
Comment (fzipi) [reply cmd/chore_copyright_update_test.go:1]: Test added.

**Review by theseion — APPROVED** [review]

**Pull request merged**